### PR TITLE
Add local-pv-provisioner

### PR DESCRIPTION
**Pull request by dulltz**
- [x] Implement local-pv-provisioner
- [x] Dockerfile
- [x] Test
- [x] CI

### CODE REVIEW

File: cke-tools/src/go.mod [hunk 1]
```diff
@@ -21,3 +21,4 @@ require (
 	sigs.k8s.io/yaml v1.1.0
 )
 
+go 1.13
```
Comment (ymmt2005) [inline cke-tools/src/go.mod]: This file is not related to this PR.

File: local-pv-provisioner/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM quay.io/cybozu/ubuntu:18.04
```
Comment (ymmt2005) [inline local-pv-provisioner/Dockerfile]: Please use `scratch`.

File: local-pv-provisioner/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          operator: In
+          values:
+          - node1
+```
+
+## Command-line flags and environment variables
+
+| Flag               | Env name              | Default              | Description                                                                                         |
+| ------------------ | --------------------- | -------------------- | --------------------------------------------------------------------------------------------------- |
+| metrics-addr       | LP_METRICS_ADDR       | `:8180`              | Bind address for the metrics endpoint.                                                              |
```
Comment (ymmt2005) [inline local-pv-provisioner/README.md]: Why do you choose a different port number from the default (8080)?
https://godoc.org/sigs.k8s.io/controller-runtime/pkg/metrics#pkg-variables

I think it should keep the default.

File: local-pv-provisioner/controllers/local_device.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	})
+	if err != nil {
+		log.Error(err, "error while walking the path", "path", dd.deviceDir)
+		return nil, err
+	}
+	return devs, nil
+}
+
+func (dd *DeviceDetector) getCapacityBytes(ctx context.Context, path string) (int64, error) {
+	out, err := exec.CommandContext(ctx, "lsblk", path, "--bytes", "--output=SIZE", "--noheadings").Output()
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/local_device.go]: The block device size can be obtained by pure Go w/o `lsblk`:
https://stackoverflow.com/a/57822606/1493661

File: cke-tools/src/go.mod (hunk 1 above)
Comment (dulltz) [reply cke-tools/src/go.mod]: OK

File: local-pv-provisioner/README.md (hunk 3 above)
Comment (dulltz) [reply local-pv-provisioner/README.md]: Agrred

File: local-pv-provisioner/controllers/local_device.go (hunk 4 above)
Comment (dulltz) [reply local-pv-provisioner/controllers/local_device.go]: I'll try it

File: local-pv-provisioner/kind/local_pv_provisioner_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to TempDir %w", err)
+	}
+
+	err = exec.Command("sudo", "mkdir", "-p", deviceSymlinkPath).Run()
+	if err != nil {
+		return nil, fmt.Errorf("failed to mkdir %w", err)
+	}
+
+	exec.Command("sudo", "mkdir", "-p", deviceSymlinkPath).Run()
```
Comment (masa213f) [inline local-pv-provisioner/kind/local_pv_provisioner_test.go]: duplicated

File: local-pv-provisioner/controllers/persistent_volume.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+			NodeAffinity: &corev1.VolumeNodeAffinity{
+				Required: &corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{
+					{MatchExpressions: []corev1.NodeSelectorRequirement{
+						{Key: hostNameLabelKey, Operator: "In", Values: []string{dd.nodeName}},
+					}},
+				}},
+			},
+		},
+	}
+	op, err := ctrl.CreateOrUpdate(ctx, dd.Client, pv, func() error {
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/persistent_volume.go]: This is a typical wrong usage of `CreateOrUpdate`. The object must be empty except for its name and namespace.
Please read the example carefully.
https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

Comment (ymmt2005) [reply local-pv-provisioner/controllers/persistent_volume.go]: In addition, I guess you do not want to update existing PVs, don't you?

File: local-pv-provisioner/controllers/persistent_volume.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+		pv.SetOwnerReferences([]v1.OwnerReference{*ownerRef})
+		return nil
+	})
+	if err != nil {
+		log.Error(err, "unable to create PV")
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/persistent_volume.go]: Why don't you return an error?

File: local-pv-provisioner/controllers/device_detector.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	nodeGVK, err := apiutil.GVKForObject(node, dd.scheme)
+	if err != nil {
+		log.Error(err, "unable to get GVK")
+		return err
+	}
+
+	var pvList corev1.PersistentVolumeList
+	if err := dd.List(ctx, &pvList, client.MatchingLabels{nodeNameLabel: dd.nodeName}); err != nil {
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/device_detector.go]: For better performance, you should create an in-memory index.

Ref:
- https://bozuman.cybozu.com/k/#/space/3801/thread/38010/2187144/2877873
- https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#setup

File: kind-node/src/vendor/gopkg.in/yaml.v2/go.sum [hunk 9]
```diff
@@ -0,0 +1 @@
+gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply kind-node/src/vendor/gopkg.in/yaml.v2/go.sum]: Removed change of go.sum under the vendor directory.

File: local-pv-provisioner/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+## What resources are created
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: node1-sda
+  ownerReferences:
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply local-pv-provisioner/README.md]: Changed to use OwnerReferences instead of Labels.

File: local-pv-provisioner/controllers/device_detector.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	log := dd.log.WithValues("node", dd.nodeName)
+
+	node := new(corev1.Node)
+	if err := dd.Get(ctx, types.NamespacedName{Name: dd.nodeName}, node); err != nil {
+		if !apierrs.IsNotFound(err) {
+			log.Error(err, "unable to fetch Node", "node", dd.nodeName)
+			return err
+		}
+	}
+
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/device_detector.go]: Changed to perform error handling regardless of the existence of Node.

File: local-pv-provisioner/controllers/device_detector.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	devices, err := dd.listLocalDevices()
+	if err != nil {
+		log.Error(err, "unable to list local devices")
+		return err
+	}
+	log.Info("local devices", "devices", devices)
+
+	nodeRef := &metav1.OwnerReference{
```
[the remark this answers is not in the document]
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/device_detector.go]: Changed to use `SetControllerReference` instead of `OwnerReference` to bind an owner reference.

File: local-pv-provisioner/controllers/persistent_volume.go (hunk 6 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/persistent_volume.go]: Change to use CreateOrUpdate correctly referred by an example.

File: local-pv-provisioner/controllers/device_detector.go (hunk 8 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/device_detector.go]: Removed listing up of PV because it is not for use.

File: local-pv-provisioner/controllers/persistent_volume.go (hunk 7 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/persistent_volume.go]: Changed to return an error and checked all error handlings.

File: local-pv-provisioner/cmd/root.go [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+func init() {
+	fs := rootCmd.Flags()
+	fs.String("metrics-addr", ":8080", "Listen address for metrics")
+	fs.Bool("development", false, "Use development logger config")
+	fs.String("device-dir", "/dev/disk/by-path/", "Path to the directory that stores the devices for which PersistentVolumes are created.")
+	fs.String("device-name-filter", ".*", "A regular expression that allows selection of devices on device-idr to be created PersistentVolume.")
+	fs.String("node-name", "", "The name of Node on which this program is running")
+	fs.Uint("polling-interval", 10, "Polling interval to check devices. It is set by a second.")
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/root.go]: [`fs.Duration`](https://godoc.org/github.com/spf13/pflag#FlagSet.Duration) is more appropriate.

File: local-pv-provisioner/cmd/run.go [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+		Scheme:             scheme,
+		MetricsBindAddress: config.metricsAddr,
+		LeaderElection:     false,
+	})
+	if err != nil {
+		log.Error(err, "unable to start manager")
+		return err
+	}
+	if len(config.nodeName) == 0 {
+		err = errors.New("node-name must not by empty")
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/run.go]:
```suggestion
		err = errors.New("node-name must not be empty")
```

File: local-pv-provisioner/cmd/run.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+		log.Error(err, "validation error")
+		return err
+	}
+	if !filepath.IsAbs(config.deviceDir) {
+		err = errors.New("device-dir is must be a absolute path")
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/run.go]:
```suggestion
		err = errors.New("device-dir must be an absolute path")
```

File: local-pv-provisioner/cmd/run.go [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		log.Error(err, "device-dir is must be a absolute path")
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/run.go]: ditto

File: local-pv-provisioner/cmd/run.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+		return err
+	}
+	info, err := os.Stat(config.deviceDir)
+	if err != nil {
+		log.Error(err, "unable to get status of divice direcotry", "device-dir", config.deviceDir)
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/run.go]:
```suggestion
		log.Error(err, "unable to get status of device direcotry", "device-dir", config.deviceDir)
```

File: local-pv-provisioner/cmd/root.go [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+	"k8s.io/klog"
+)
+
+var config struct {
+	metricsAddr      string
+	nodeName         string
+	deviceDir        string
+	deviceNameFilter string
+	development      bool
+	pollingInterval  int
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/root.go]: should be a `time.Duration` as suggested below.

File: local-pv-provisioner/controllers/device_detector.go [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+				return err
+			}
+		case <-ch:
+			return nil
+		}
+	}
+}
+
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;update;patch;watch
+// +kubebuilder:rbac:groups="",resources=nodes/status,verbs=get;update;patch
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/device_detector.go]: Are Nodes really to be modified?

File: local-pv-provisioner/controllers/persistent_volume.go [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	storageClass     = "local-storage"
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/persistent_volume.go]: This storage class name is quite important and an external specification
to bind the persistent volumes created by `local-pv-provisioner` through
persistent volume claims (PVC).

I'd like to export this as a public constant.

File: local-pv-provisioner/controllers/persistent_volume.go [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+	hostNameLabelKey = "kubernetes.io/hostname"
+)
+
+func (dd *DeviceDetector) pvName(devName string) string {
+	hasher := sha1.New()
+	hasher.Write([]byte(devName))
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/persistent_volume.go]: Use `sha1.Sum` for better performance.
https://golang.org/pkg/crypto/sha1/#Sum

But why don't you use `devName` as is?

File: local-pv-provisioner/controllers/device_detector.go [hunk 22]
```diff
@@ -80,10 +80,8 @@ func (dd *DeviceDetector) Start(ch <-chan struct{}) error {
 	}
 }
 
-// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;update;patch;watch
-// +kubebuilder:rbac:groups="",resources=nodes/status,verbs=get;update;patch
-// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get;list;create;update;patch;delete;watch
-// +kubebuilder:rbac:groups="",resources=persistentvolumes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=list;create;update;watch
```
Comment (ymmt2005) [inline local-pv-provisioner/controllers/device_detector.go]: Please use a set of `get,list,watch` for read access and `update,patch` for write access.

File: local-pv-provisioner/cmd/root.go (hunk 13 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/root.go]: Changed to use Duration.

File: local-pv-provisioner/cmd/run.go (hunk 14 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/run.go]: Fixed

File: local-pv-provisioner/cmd/run.go (hunk 15 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/run.go]: Fixed

File: local-pv-provisioner/cmd/run.go (hunk 16 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/run.go]: Fixed

File: local-pv-provisioner/cmd/run.go (hunk 17 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/run.go]: Fixed

File: local-pv-provisioner/cmd/root.go (hunk 18 above)
Comment (llamerada-jp) [reply local-pv-provisioner/cmd/root.go]: Changed to use Duration.

File: local-pv-provisioner/controllers/device_detector.go (hunk 19 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/device_detector.go]: Checked and updated RBAC configuration

File: local-pv-provisioner/controllers/persistent_volume.go (hunk 20 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/persistent_volume.go]: Changed to public constant.

File: local-pv-provisioner/controllers/persistent_volume.go (hunk 21 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/persistent_volume.go]: Removed sha1 because devName's rule was changed.

File: local-pv-provisioner/controllers/device_detector.go (hunk 22 above)
Comment (llamerada-jp) [reply local-pv-provisioner/controllers/device_detector.go]: Updated RBAC rules.

File: local-pv-provisioner/Makefile [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+
+docker: build
+	docker build . -t $(IMAGE_TAG)
+
+setup:
+	curl -sfL https://go.kubebuilder.io/dl/$(KUBEBUILDER_VERSION)/$(GOOS)/$(GOARCH) | tar -xz -C /tmp/
+	$(SUDO) rm -rf /usr/local/kubebuilder
+	$(SUDO) mv /tmp/kubebuilder_$(KUBEBUILDER_VERSION)_$(GOOS)_$(GOARCH) /usr/local/kubebuilder
+	cd /tmp; GO111MODULE=on GOFLAGS= go get sigs.k8s.io/controller-tools/cmd/controller-gen@v$(CTRLTOOLS_VERSION)
+	GO111MODULE=on GOFLAGS= go install github.com/onsi/ginkgo/ginkgo
```
Comment (ymmt2005) [inline local-pv-provisioner/Makefile]: `ginkgo` command is not used, is it?
```suggestion
```

File: local-pv-provisioner/README.md [hunk 24]
```diff
@@ -0,0 +1,91 @@
+[![Docker Repository on Quay](https://quay.io/repository/cybozu/local-pv-provisioner/status "Docker Repository on Quay")](https://quay.io/repository/cybozu/local-pv-provisioner)
+
+local-pv-provisioner
+====================
+
+`local-pv-provisioner` is a custom controller that creates [local](https://kubernetes.io/docs/concepts/storage/volumes/#local) PersistentVolume(PV) resources from devices that match the specified conditions.
+
+* The PVs are linked to a node by `ownerReferences` setting.
+* The PVs will be removed along with the deletion of the node.
```
Comment (ymmt2005) [inline local-pv-provisioner/README.md]: These two lines say the same thing.

File: local-pv-provisioner/README.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+          values:
+          - node1
+```
+
+* `metadata.name` is decided according to the following rules.
+   * The name is a concatenation of `local`, node name, and device name with `-`.
+   * If it contains characters other than alphabets, numbers, `-` and `.`, it is replaced with `-`.
+* `spec.storageClassName` is automatically set a value `local-storage`.
+* `spec.capacity.storage` is decided from the max capacity of the device.
+
```
Comment (ymmt2005) [inline local-pv-provisioner/README.md]: Please add a section to explain how to bind the created PV with a PVC.

File: local-pv-provisioner/cmd/root.go [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func init() {
+	fs := rootCmd.Flags()
+	fs.String("metrics-addr", ":8080", "Listen address for metrics")
+	fs.Bool("development", false, "Use development logger config")
+	fs.String("device-dir", "/dev/disk/by-path/", "Path to the directory that stores the devices for which PersistentVolumes are created.")
+	fs.String("device-name-filter", ".*", "A regular expression that allows selection of devices on device-idr to be created PersistentVolume.")
+	fs.String("node-name", "", "The name of Node on which this program is running")
+	fs.Duration("polling-interval", 10*time.Second, "Polling interval to check devices.")
```
Comment (ymmt2005) [inline local-pv-provisioner/cmd/root.go]: The default interval 10 second seems too short.
It can be, for example, 5 minutes.

File: local-pv-provisioner/Makefile (hunk 23 above)
Comment (masa213f) [reply local-pv-provisioner/Makefile]: Yes. `ginkgo` command was not used. I deleted it.

File: local-pv-provisioner/README.md (hunk 24 above)
Comment (masa213f) [reply local-pv-provisioner/README.md]: I fixed them to one line.

File: local-pv-provisioner/README.md (hunk 25 above)
Comment (masa213f) [reply local-pv-provisioner/README.md]: I added a Usage section.

File: local-pv-provisioner/cmd/root.go (hunk 26 above)
Comment (masa213f) [reply local-pv-provisioner/cmd/root.go]: I changed it.

File: local-pv-provisioner/README.md [hunk 27]
```diff
@@ -0,0 +1,175 @@
...
+          volumeDevices:
+            - name: sample-volume
+              devicePath: /dev/sample-dev
+      volumes:
+        - name: sample-volume
+          persistentVolumeClaim:
+            claimName: sample-pvc
+    ```
+
+3. After the pod was created, the PVC will be bind to a PV.
```
Comment (ymmt2005) [inline local-pv-provisioner/README.md]:
```suggestion
3. After the pod was created, the PVC will be bound to a PV.
```

**Comment by ymmt2005** [comment]
@masa213f should you approve the PR?  thanks.
